<a href="https://colab.research.google.com/github/dazzens/dazzens/blob/main/EDA_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

------------------
---------
-----

# **Пет проект EDA (исследовательский анализ данных).**




Автор: А.Г.Х.

Начало: 22.06.2026


Окончание: --.--.2026

----



##**Бизнес-контекст**

- Интернет-магазин продает товары для дома и декора по всему миру. За последний год ассортимент вырос на 30%, но темпы роста выручки замедлились. Необходимо исследовать данные о продажах, чтобы понять, какие товары и страны приносят основную прибыль, кто самые ценные клиенты и как оптимизировать ценовую политику.

-----

# **Цель исследования**

- Выявить ключевые факторы, влияющие на динамику продаж интернет-магазина, для разработки рекомендаций по увеличению выручки и удержанию клиентов.


##**Вопросы:**


- 1. Как менялись продажи по месяцам? Динамика выручки, количества заказов и сезонность.


- 2. Какие товары самые популярные? Топ по количеству продаж, топ по выручке, топ по частоте заказов.

- 3. Какой средний чек и размер корзины? Среднее количество товаров в заказе, средняя сумма заказа

- 4. Какие страны приносят больше всего выручки? Топ по выручке, топ по количеству заказов, топ по среднему чеку

- 5. Кто самые прибыльные клиенты? Топ клиентов по сумме трат, по частоте покупок, анализ повторных покупок

- 6. Как цена влияет на продажи? Распределение цен, зависимость количества продаж от цены, оптимальные ценовые сегменты

- 7. Какие товары, страны и клиентские сегменты вносят наибольший вклад в выручку?


## **Содержание проекта**



1.  **Первичный осмотр и подготовка данных**
    *   Импорт библиотек и загрузка данных.
    *   Предварительный анализ структуры данных (типы, объем, пропуски).
    *   Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
    *   Проверка и обработка пропусков (`CustomerNo`).
    *   Проверка дубликатов и выбросов (`Price`, `Quantity`).

2.  **Предварительный анализ и создание новых признаков**
    *   Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
    *   Создание признаков: месяц, день недели, год из даты.
    *   Создание признака: общая сумма заказа (`Price` x `Quantity`).
    *   Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

3.  **Исследовательский анализ данных (EDA)**
    *   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
    *   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
    *   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
    *   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
    *   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
    *   Анализ ценовых сегментов: зависимость количества продаж от цены.

4.  **Ключевые выводы и рекомендации**
    *   Формирование основных инсайтов по каждому блоку анализа.
    *   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.


    

## **Описание данных**

**Транзакционные данные (онлайн-магазин):**

| | | |
|------|----------|------------|
| `TransactionNo` | Уникальный идентификатор транзакции/заказа | `object` |
| `Date` | Дата совершения покупки | `object` |
| `ProductNo` | Уникальный идентификатор товара | `object` |
| `ProductName` | Название товара | `object` |
| `Price` | Цена за единицу товара | `float64` |
| `Quantity` | Количество единиц товара в заказе | `int64` |
| `CustomerNo` | Уникальный идентификатор клиента | `float64` |
| `Country` | Страна клиента | `object` |

---

**Характеристики данных:**

|  |  |
|------------|----------|
| **Объем данных** | 536 350 записей |
| **Количество клиентов** | 4 738 |
| **Количество уникальных товаров** | 3 768 |
| **Количество стран** | 38 |
| **Пропуски** | 55 записей без `CustomerNo` (0.01%) |



----
----

# **Этап 1 Первичный осмотр и подготовка данных**

- Импорт библиотек и загрузка данных.
- Предварительный анализ структуры данных (типы, объем, пропуски).
- Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
- Проверка и обработка пропусков (`CustomerNo`).
- Проверка дубликатов и выбросов (`Price`, `Quantity`).



## **Шаг 1. Загрузка данных**

- Импорт библиотек и загрузка датасета из внешнего источника.

---
### **Библиотеки**




In [1]:

#Необходимые библиотеки

try:
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    #from scipy.stats import ttest_ind
    #from matplotlib_venn import venn2
    import kagglehub
    import os
    print("Все библиотеки загружены")

except ImportError:
    !pip install numpy pandas matplotlib seaborn   kagglehub -q
    #scipy matplotlib-venn
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    #from scipy.stats import ttest_ind
    #from matplotlib_venn import venn2
    import kagglehub
    import os
    print("Библиотеки загружены через pip")



Все библиотеки загружены


### **Загрузка датасета**

In [2]:
#Загрузка датасета
path = kagglehub.dataset_download("gabrielramos87/an-online-shop-business")
file_path = os.path.join(path, [x for x in os.listdir(path) if x.endswith('.csv')][0])

#Сохранение
df = pd.read_csv(file_path)



Using Colab cache for faster access to the 'an-online-shop-business' dataset.


**Итого:**

 - Установлены необходимые библиотеки и датасет сохранён в `df`

-----
----
----

## **Шаг 2. Общий осмотр данных**

Первичный анализ структуры данных включает оценку следующих параметров, таких как:

- размерность датасета
- типы признаков и их соответствие содержанию
- наличие пропусков и их доля в каждом поле
- основные статистические характеристики числовых переменных

In [3]:
# Основная информация
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 536350 entries, 0 to 536349
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  536350 non-null  object 
 1   Date           536350 non-null  object 
 2   ProductNo      536350 non-null  object 
 3   ProductName    536350 non-null  object 
 4   Price          536350 non-null  float64
 5   Quantity       536350 non-null  int64  
 6   CustomerNo     536295 non-null  float64
 7   Country        536350 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.7+ MB


In [4]:
#Вывод первых 10 строк
df.head(10)


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
0,581482,12/9/2019,22485,Set Of 2 Wooden Market Crates,21.47,12,17490.0,United Kingdom
1,581475,12/9/2019,22596,Christmas Star Wish List Chalkboard,10.65,36,13069.0,United Kingdom
2,581475,12/9/2019,23235,Storage Tin Vintage Leaf,11.53,12,13069.0,United Kingdom
3,581475,12/9/2019,23272,Tree T-Light Holder Willie Winkie,10.65,12,13069.0,United Kingdom
4,581475,12/9/2019,23239,Set Of 4 Knick Knack Tins Poppies,11.94,6,13069.0,United Kingdom
5,581475,12/9/2019,21705,Bag 500g Swirly Marbles,10.65,24,13069.0,United Kingdom
6,581475,12/9/2019,22118,Joy Wooden Block Letters,11.53,18,13069.0,United Kingdom
7,581475,12/9/2019,22119,Peace Wooden Block Letters,12.25,12,13069.0,United Kingdom
8,581475,12/9/2019,22217,T-Light Holder Hanging Lace,10.65,12,13069.0,United Kingdom
9,581475,12/9/2019,22216,T-Light Holder White Lace,10.55,24,13069.0,United Kingdom


In [5]:
#Осмотр случайных строк
df.sample(10)


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
128074,572304,10/23/2019,21171,Bathroom Metal Sign,10.81,25,17131.0,United Kingdom
229922,564172,8/23/2019,23177,Treasure Island Book Box,12.56,3,16033.0,United Kingdom
415458,546896,3/17/2019,22478,Birdhouse Garden Marker,12.77,1,15896.0,United Kingdom
483516,540779,1/11/2019,21703,Bag 125g Swirly Marbles,10.68,12,15346.0,United Kingdom
50359,577985,11/22/2019,21537,Red Retrospot Pudding Bowl,14.61,4,15116.0,United Kingdom
475612,541280,1/17/2019,84029G,Knitted Union Flag Hot Water Bottle,14.09,4,15296.0,United Kingdom
90676,574950,11/8/2019,46776F,Woven Rose Garden Cushion Cover,14.48,1,15950.0,United Kingdom
258349,561394,7/27/2019,23174,Regency Sugar Bowl Green,14.50,6,17315.0,United Kingdom
300646,557858,6/23/2019,22900,Set 2 Tea Towels I Love London,13.58,1,14159.0,United Kingdom
519086,537854,12/8/2018,18097C,White Tall Porcelain T-Light Holder,12.86,24,13984.0,United Kingdom


In [6]:
#Осмотр по различным показателям
df.describe()


,Price,Quantity,CustomerNo
count,536350.000000,536350.000000,536295.000000
mean,12.662182,9.919347,15227.893178
std,8.490450,216.662300,1716.582932
min,5.130000,-80995.000000,12004.000000
25%,10.990000,1.000000,13807.000000
50%,11.940000,3.000000,15152.000000
75%,14.090000,10.000000,16729.000000
max,660.620000,80995.000000,18287.000000


### **Результаты первичного осмотра**

**Общая информация:**
- Датасет содержит 536 350 записей, 8 столбцов.
- Пропуски: в поле `CustomerNo`.
- Типы данных: `Date` и `CustomerNo` требуют преобразования для дальнейшего анализа.
- `Price`: среднее — 12.66 руб., медиана — 11.94 руб. Диапазон: от 5.13 до 660.62 руб.
- `Quantity`: среднее — 9.92 шт., медиана — 3 шт. Диапазон: от -80 995 до 80 995 шт., что указывает на наличие экстремальных значений.
  
**Выводы:**
  
  -  Данные содержат аномалии в `Quantity` (отрицательные и экстремально высокие значения) и выбросы в `Price`.
  - Требуется преобразование типов и принятие решения по обработке аномалий. В остальном данные пригодны для анализа.
  - Требуется осомотр пропусков в поле `CustomerNo`.


## **Шаг 3. Обработка типов данных**

- проверка и приведение типов данных к форматам, соответствующим их содержанию
- обеспечение согласованности типов для последующих операций
- оптимизация структуры данных для дальнейшей работы
- `Date` преобразовать в `datetime` для временной аналитики
- `CustomerNo` привести  к `Int64` для корректной группировки клиентов

In [7]:
#Осмотр типов данных

df.dtypes

,0
TransactionNo,object
Date,object
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,float64
Country,object


In [8]:
# преобразование данных
df['Date'] = pd.to_datetime(df['Date'])
df['CustomerNo'] = df['CustomerNo'].astype('Int64')

#проверка
df.dtypes

,0
TransactionNo,object
Date,datetime64[ns]
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,Int64
Country,object


### **Итог преобразования**

Типы данных приведены к форматам, необходимым для дальнейшего анализа

- `Date` преобразована в `datetime`, что открывает доступ к исследованию временных рядов и сезонных срезов;
- `CustomerNo` переведён в `Int64` с поддержкой пропусков, что обеспечивает корректную идентификацию клиентов при группировках.

Остальные типы данных оставлены без изменений.

---
----

## **Шаг 4 Поиск и обработка пропусков**

- Проверка данных на наличие пропущенных значений
- оценка их объема
- анализ возможного влияния на результаты исследования и принятие
- решения по их обработке

In [9]:
# колличесвто и доля пропусков

df.isnull().sum(), df.isnull().mean()

(TransactionNo     0
 Date              0
 ProductNo         0
 ProductName       0
 Price             0
 Quantity          0
 CustomerNo       55
 Country           0
 dtype: int64,
 TransactionNo    0.000000
 Date             0.000000
 ProductNo        0.000000
 ProductName      0.000000
 Price            0.000000
 Quantity         0.000000
 CustomerNo       0.000103
 Country          0.000000
 dtype: float64)

In [10]:
# отображение данных


df[df['CustomerNo'].isnull()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,<NA>,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,<NA>,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,<NA>,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,<NA>,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,<NA>,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,<NA>,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,<NA>,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,<NA>,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,<NA>,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,<NA>,United Kingdom


####**Описание и дальнейшие действия:**

* Пропуски обнаружены только в поле `CustomerNo` и составляют **55** записей (0.01%).
* Для сохранения всех наблюдений пропуски будут заменены значением `-1`, которое в дальнейшем будет обозначать неизвестного клиента

----

In [11]:

# Замена NaN на специальное значение
df['CustomerNo'] = df['CustomerNo'].fillna(-1)

# проверка после заполнения пропусков
display(df[df['CustomerNo'] == -1].head(10))

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,-1,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,-1,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,-1,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,-1,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,-1,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,-1,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,-1,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,-1,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,-1,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,-1,United Kingdom


### **Итог работы c пропусками:**


Пропуски в `CustomerNo` обработаны заменой на значение `-1`.


Пропущенных значений в датасете не осталось.



-----
-----

---
---

## **Шаг 5. Поиск и анализ  явных/неявных дубликатов**

- Выявление избыточных записей (полностью совпадающих строк)

- оценка их объёма и доли в общем массиве данных

- анализ возможного влияния дубликатов на агрегированные метрики

- принятие решения по обработке: удаление или агрегация

---

###**Анализ явных дубликатов:**

- поиск полностью совпадающих строк по всем колонкам
- оценка количества и доли дубликатов в данных
- анализ влияния на метрики (выручка, количество заказов, товаров)


In [13]:
# Количество и доля явных дубликатов

duplicates_count = df.duplicated().sum()
duplicates_percent = (duplicates_count / len(df)) * 100

print(f"Количество дубликатов: {duplicates_count}")
print(f"Процент дубликатов: {duplicates_percent:.2f}%")

Количество дубликатов: 5163
Процент дубликатов: 0.98%


In [14]:
#Вывод записей дубликатов

df[df.duplicated()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
985,581497,2019-12-09,21481,Fawn Blue Hot Water Bottle,7.24,1,17497,United Kingdom
1365,581538,2019-12-09,23275,Set Of 3 Hanging Owls Ollie Beak,6.19,1,14446,United Kingdom
1401,581538,2019-12-09,22992,Revolver Wooden Ruler,6.19,1,14446,United Kingdom
1406,581538,2019-12-09,22694,Wicker Star,6.19,1,14446,United Kingdom
1409,581538,2019-12-09,23343,Jumbo Bag Vintage Christmas,6.19,1,14446,United Kingdom
3255,581237,2019-12-08,23498,Classic Bicycle Clips,6.19,1,15443,United Kingdom
3259,581237,2019-12-08,23497,Classic Chrome Bicycle Bell,6.19,1,15443,United Kingdom
3261,581237,2019-12-08,23509,Mini Playing Cards Fun Fair,6.19,2,15443,United Kingdom
3284,581237,2019-12-08,23378,Pack Of 12 50'S Christmas Tissues,6.19,1,15443,United Kingdom
3593,581253,2019-12-08,23427,Stool Home Sweet Home,6.19,1,16891,United Kingdom


####**Дальнейшие действия с явными дубликатами:**


Обнаружено **5200** дубликатов (0.97% от всех записей).  

Удаление не повлияет на общую структуру данных и исключит искажение метрик.

In [15]:
#Удаление дубликатов и проверка

df = df.drop_duplicates()

df.duplicated().sum()

np.int64(0)

####**Итог обработки:**

Явные дубликаты полностью удалены.

 В датасете не осталось полностью совпадающих строк.


----
-----

### **Анализ неявных дубликатов:**

- Проверка наличия повторяющихся товаров в рамках одной транзакции.

- Выявление причин возникновения (технические дубли, добавление товара частями, ошибки ввода).

- Оценка влияния на количественные метрики (общее количество товаров в заказе, сумма выручки).

- Выбор стратегии обработки: агрегация (суммирование Quantity) или оставление как есть.

- Обоснование принятого решения для обеспечения целостности данных.

In [16]:
# Поиск неявных дубликатов

implicit_duplicates = (
    df.groupby(['TransactionNo', 'ProductNo'])
      .size()
      .reset_index(name='count')
      .query('count > 1')
)

print(f'Найдено неявных дубликатов: {len(implicit_duplicates)}')
display(implicit_duplicates.head(10))




Найдено неявных дубликатов: 5523


,TransactionNo,ProductNo,count
133,536381,71270,2
515,536409,85116,2
518,536409,90199C,3
529,536412,21448,2
535,536412,21738,3
538,536412,22077,3
541,536412,22144,2
545,536412,22243,3
547,536412,22273,2
562,536412,22749,2


In [17]:
# Пример заказа с повторяющимися товарами
sample_transaction = implicit_duplicates['TransactionNo'].iat[0]

display(
    df.loc[
        df['TransactionNo'] == sample_transaction,
        ['TransactionNo', 'Date', 'ProductNo', 'ProductName', 'Quantity']
    ].sort_values('ProductNo')
)


,TransactionNo,Date,ProductNo,ProductName,Quantity
533393,536381,2018-12-01,15056N,Edwardian Parasol Natural,2
533379,536381,2018-12-01,21166,Cook With Wine Metal Sign,1
533378,536381,2018-12-01,21169,You're Confusing Me Metal Sign,3
533380,536381,2018-12-01,21175,Gin And Tonic Diet Metal Sign,2
533400,536381,2018-12-01,21523,Doormat Fancy Font Home Sweet Home,10
533391,536381,2018-12-01,21533,Retrospot Large Milk Jug,1
533392,536381,2018-12-01,21557,Set Of 6 Funky Beakers,2
533371,536381,2018-12-01,21672,White Spot Red Ceramic Drawer Knob,6
533397,536381,2018-12-01,21731,Red Toadstool Led Night Light,2
533377,536381,2018-12-01,21934,Skull Shoulder Bag,10


In [18]:
# Формирование датафрейма для дальнейшей обработки
duplicate_rows = df.merge(
    implicit_duplicates,
    on=['TransactionNo', 'ProductNo'],
    how='inner'
)

----
**// Анализ отрицательного количества**

In [19]:
# Анализ отрицательных неявных дубликатов

# Выделяем отрицательные неявные дубликаты
negative_duplicates = duplicate_rows[duplicate_rows['Quantity'] < 0]

print(f"Отрицательных неявных дубликатов: {len(negative_duplicates)}")




Отрицательных неявных дубликатов: 0


In [20]:
# Объединяем повторяющиеся позиции
negative_grouped = (
    negative_duplicates
    .groupby(
        ['TransactionNo', 'ProductNo', 'Date', 'ProductName',
         'Price', 'CustomerNo', 'Country'],
        as_index=False
    )['Quantity']
    .sum()
)

print(f"До объединения: {len(negative_duplicates)}")
print(f"После объединения: {len(negative_grouped)}")


До объединения: 0
После объединения: 0


In [21]:
# Проверяем, возникают ли разные цены у одного товара в одной транзакции
price_conflicts = (
    negative_grouped
    .groupby(['TransactionNo', 'ProductNo'])['Price']
    .nunique()
)

print(f"Комбинаций с разными ценами: {(price_conflicts > 1).sum()}")



Комбинаций с разными ценами: 0


In [22]:
# Транзакции с разными ценами
price_diff = (
    negative_grouped
    .groupby(['TransactionNo', 'ProductNo'])['Price']
    .nunique()
    .reset_index(name='unique_prices')
    .query('unique_prices > 1')
)

display(price_diff)

# Детальный просмотр этих случаев
display(
    negative_grouped.merge(
        price_diff[['TransactionNo', 'ProductNo']],
        on=['TransactionNo', 'ProductNo']
    ).sort_values(['TransactionNo', 'ProductNo', 'Price'])
)

,TransactionNo,ProductNo,unique_prices


,TransactionNo,ProductNo,Date,ProductName,Price,CustomerNo,Country,Quantity


- В ходе проверки выявлено две комбинации TransactionNo–ProductNo, в которых один и тот же товар имеет разные значения Price. Это свидетельствует о том, что объединение только по номеру транзакции и товара может привести к потере информации. Поэтому поле Price сохранено в ключе группировки, а такие записи не объединялись.

In [23]:
# Пример результата после объединения
display(negative_grouped.head(10))

,TransactionNo,ProductNo,Date,ProductName,Price,CustomerNo,Country,Quantity


#### **Обработка неявных дубликатов: Объединение позиций**

Неявные дубликаты, которые представляют собой повторные записи одного и того же товара в рамках одной транзакции, будут объединены. Для этого мы сгруппируем данные по `TransactionNo`, `Date`, `ProductNo`, `ProductName`, `Price`, `CustomerNo` и `Country`, а затем просуммируем `Quantity`. Это позволит сохранить общую информацию о количестве проданного товара в рамках транзакции, избегая потери данных.

In [24]:
# Объединяем неявные дубликаты, суммируя Quantity
df_cleaned = df.groupby(['TransactionNo', 'Date', 'ProductNo', 'ProductName', 'Price', 'CustomerNo', 'Country'], as_index=False)['Quantity'].sum()

print(f"Размер датафрейма до обработки неявных дубликатов: {df.shape}")
print(f"Размер датафрейма после обработки неявных дубликатов: {df_cleaned.shape}")

display(df_cleaned.head())

Размер датафрейма до обработки неявных дубликатов: (522602, 8)
Размер датафрейма после обработки неявных дубликатов: (517898, 8)


,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
0,536365,2018-12-01,21730,Glass Star Frosted T-Light Holder,14.61,17850,United Kingdom,6
1,536365,2018-12-01,22752,Set 7 Babushka Nesting Boxes,18.09,17850,United Kingdom,2
2,536365,2018-12-01,71053,White Moroccan Metal Lantern,13.72,17850,United Kingdom,6
3,536365,2018-12-01,84029E,Red Woolly Hottie White Heart,13.72,17850,United Kingdom,6
4,536365,2018-12-01,84029G,Knitted Union Flag Hot Water Bottle,13.72,17850,United Kingdom,6


### **Обший вывод по  по обработке явных и неявных дубликатов:**


- Неявные дубликаты были успешно объединены, в результате чего размер датафрейма уменьшился с `531150` до `525367` строк. Это означает, что `5783` строк, представляющих собой повторные записи одних и тех же товаров в рамках одной транзакции, были корректно объединены, а их количества просуммированы. Теперь каждая строка в датафрейме `df_cleaned` представляет уникальную позицию товара в каждой транзакции с корректной общей `Quantity`.

-----
----
---

#!// **Анализ отриц значений**






### **Анализ отрицательных значений Quantity**

Отрицательные значения в столбце `Quantity` (обнаружено `8548` записей, составляющих `1.61%` от общего числа) обычно указывают на возвраты товаров или отмененные заказы. Для целей анализа продаж и выручки, эти записи могут искажать общую картину, так как они представляют собой не продажи, а их отмену.

**Предлагаемые действия:**

1.  **Выделить возвраты**: Если `TransactionNo` начинается с 'C', это может прямо указывать на отмену (cancellation). Необходимо проверить это предположение.
2.  **Удалить отрицательные значения**: Для анализа чистых продаж, возвраты часто исключают из основного датафрейма. Их можно анализировать отдельно, если требуется изучить динамику возвратов.
3.  **Преобразовать в абсолютные значения**: Если цель — понять общий объем движения товаров, можно использовать абсолютные значения, но это не рекомендуется для анализа выручки.

В рамках данного EDA, мы будем считать, что отрицательные `Quantity` представляют собой возвраты/отмены и **удалим их из основного датафрейма**, чтобы сфокусироваться на чистых продажах.

In [12]:
# Проверяем транзакции, начинающиеся с 'C', которые могут быть отменами
cancellation_transactions = df[df['Quantity'] < 0]['TransactionNo'].apply(lambda x: x.startswith('C')).sum()
print(f"Количество транзакций с отрицательным Quantity, начинающихся с 'C': {cancellation_transactions}")

# Удаляем строки с отрицательным значением Quantity
df_filtered = df[df['Quantity'] > 0].copy()

print(f"Размер датафрейма до удаления отрицательных Quantity: {df.shape}")
print(f"Размер датафрейма после удаления отрицательных Quantity: {df_filtered.shape}")

# Обновляем основной DataFrame
df = df_filtered.copy()

Количество транзакций с отрицательным Quantity, начинающихся с 'C': 8585
Размер датафрейма до удаления отрицательных Quantity: (536350, 8)
Размер датафрейма после удаления отрицательных Quantity: (527765, 8)


----
----
---

## **Шаг 6. Детальный осмотр данных:**

In [25]:
display(df[df['Quantity'] < 0].head(15).sort_values(by='Quantity'))

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country


In [26]:
display(df[df['Price'] > 27].head(15).sort_values(by='Price', ascending=False))

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
49340,578042,2019-11-22,22826,Love Seat Antique White Metal,97.37,6,17500,United Kingdom
12297,580730,2019-12-05,22847,Bread Bin Diner Style Ivory,44.37,1,15730,United Kingdom
12296,580730,2019-12-05,22846,Bread Bin Diner Style Red,44.37,1,15730,United Kingdom
49342,578065,2019-11-22,22423,Regency Cakestand 3 Tier,35.83,2,15065,United Kingdom
49343,578065,2019-11-22,22622,Box Of Vintage Alphabet Blocks,35.83,1,15065,United Kingdom
49341,577844,2019-11-22,22423,Regency Cakestand 3 Tier,35.83,13,16844,United Kingdom
49344,578067,2019-11-22,22423,Regency Cakestand 3 Tier,35.83,3,12067,United Kingdom
49345,578067,2019-11-22,23553,Landmark Frame Camden Town,35.83,1,12067,United Kingdom
12298,580730,2019-12-05,23173,Regency Teapot Roses,30.71,2,15730,United Kingdom
12300,580730,2019-12-05,22626,Black Kitchen Scales,27.30,1,15730,United Kingdom


In [27]:
# промежуточный анализ


sum_quantity_min = (df['Quantity'] < 0).sum()
count_quantity_min = (df['Quantity'] < 0).count()

print(sum_quantity_min)
print(count_quantity_min)

sum_negative = df[df['Quantity'] < 0]['Quantity'].sum()
print(f"Сумма отрицательных Quantity: {sum_negative}")



######
#подсчет отрицательных колонок


negative_count = (df['Quantity'] < 0).sum()
print(f"Отрицательных Quantity: {negative_count:}")

# Доля отрицательных Quantity в процентах
negative_pct = (df['Quantity'] < 0).mean() * 100
print(f"Доля отрицательных Quantity: {negative_pct:.2f}%")

0
522602
Сумма отрицательных Quantity: 0
Отрицательных Quantity: 0
Доля отрицательных Quantity: 0.00%


//////

### **Осмотр уникальных позиций**

In [28]:
# Быстрый взгляд на все категориальные поля
df[['Country',
    'CustomerNo',
    'ProductName',
    'TransactionNo']].nunique().sort_values(ascending=False)


,0
TransactionNo,19790
CustomerNo,4719
ProductName,3753
Country,38


### **Анализ хвостов и подсчет хваостов**

- нужно подсчитать что выше 99 квартиля и какая сумма формируется от общего число и понять выброс это или нет

In [29]:
#Анализ хвоста цены

df['Price'].quantile([0.25,
                      0.5,
                      0.75,
                      0.9,
                      0.95,
                      0.99])




,Price
0.25,10.99
0.50,11.94
0.75,14.09
0.90,17.37
0.95,19.42
0.99,27.30


In [30]:
df['Quantity'].quantile([0.25,
                      0.5,
                      0.75,
                      0.9,
                      0.95,
                      0.99])

,Quantity
0.25,1.0
0.50,4.0
0.75,12.0
0.90,24.0
0.95,30.0
0.99,100.0


### **Анализ высоких значений Price**

Ранее мы заметили аномально высокие значения в столбце `Price` (до `660.62` руб.), что может влиять на средние показатели и искажать анализ прибыльности. Мы уже проанализировали 99-й перцентиль (`27.30`), и обнаружили, что есть значения гораздо выше этого. Эти высокие цены могут принадлежать к товарам класса люкс, редким предметам или быть ошибками ввода данных.

**Предлагаемые действия:**

1.  **Исследовать товары с аномально высокими ценами**: Проверить, что это за товары и соответствуют ли их названия таким высоким ценам. Возможно, это ошибки в данных или особые категории товаров.
2.  **Принять решение о сохранении/удалении**: Если это валидные дорогие товары, их можно оставить. Если это ошибки, их следует скорректировать или удалить. Для большинства анализов, такие выбросы могут быть проблемными.

На данный момент, мы не будем удалять эти строки, так как они могут представлять ценные данные о дорогих товарах. Вместо этого, мы признаем их существование и будем учитывать при дальнейшем анализе. Если потребуется более глубокий анализ ценовых сегментов, можно будет вернуться к их обработке.

In [31]:
# Выводим топ-10 самых дорогих товаров для анализа
display(df.sort_values(by='Price', ascending=False).head(10)[['ProductName', 'Price', 'Quantity', 'TransactionNo', 'Country']])

# Define upper_99 and high_outliers within this cell for correct execution
upper_99 = df['Price'].quantile(0.99);
high_outliers = df[df['Price'] > upper_99]

# Проверим количество уникальных TransactionNo для этих дорогих товаров
print(f"Количество уникальных транзакций для товаров с ценой выше 99-го перцентиля: {high_outliers['TransactionNo'].nunique()}")
print(f"Количество уникальных продуктов для товаров с ценой выше 99-го перцентиля: {high_outliers['ProductNo'].nunique()}")

,ProductName,Price,Quantity,TransactionNo,Country
464161,Vintage Red Kitchen Cabinet,660.62,2,542231,United Kingdom
465442,Vintage Red Kitchen Cabinet,660.62,2,542122,United Kingdom
332793,Vintage Red Kitchen Cabinet,660.62,1,554836,United Kingdom
403361,Vintage Blue Kitchen Cabinet,660.62,1,547814,United Kingdom
475341,Vintage Red Kitchen Cabinet,660.62,2,541282,United Kingdom
452783,Vintage Blue Kitchen Cabinet,660.62,1,543370,Cyprus
454316,Vintage Red Kitchen Cabinet,660.62,1,543253,United Kingdom
465441,Vintage Red Kitchen Cabinet,660.62,2,542102,Singapore
531163,Vintage Red Kitchen Cabinet,660.62,1,536835,United Kingdom
502409,Vintage Red Kitchen Cabinet,660.62,1,539080,United Kingdom


Количество уникальных транзакций для товаров с ценой выше 99-го перцентиля: 2079
Количество уникальных продуктов для товаров с ценой выше 99-го перцентиля: 228


In [32]:
#Анализ 99+ процентиля

upper_99 = df['Price'].quantile(0.99);
high_outliers = df[df['Price'] > upper_99]


display(high_outliers.sort_values(by='Price', ascending=False))

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
502409,539080,2018-12-16,22655,Vintage Red Kitchen Cabinet,660.62,1,16607,United Kingdom
504194,538999,2018-12-15,22655,Vintage Red Kitchen Cabinet,660.62,2,16003,United Kingdom
497061,539581,2018-12-20,22656,Vintage Blue Kitchen Cabinet,660.62,2,16143,United Kingdom
507610,538662,2018-12-13,22656,Vintage Blue Kitchen Cabinet,660.62,2,15159,United Kingdom
507609,538662,2018-12-13,22655,Vintage Red Kitchen Cabinet,660.62,2,15159,United Kingdom
...,...,...,...,...,...,...,...,...
84183,575718,2019-11-10,23486,Antique Heart Shelf Unit,27.32,1,16907,United Kingdom
96235,574588,2019-11-06,23486,Antique Heart Shelf Unit,27.32,2,14936,Channel Islands
96236,574661,2019-11-06,23486,Antique Heart Shelf Unit,27.32,1,13447,United Kingdom
70141,576511,2019-11-15,23486,Antique Heart Shelf Unit,27.32,1,13846,United Kingdom


### **Отрицательный показатель количества**

### **Причина и обработка отрицательных значений Quantity**

Отрицательные значения в столбце `Quantity` обычно указывают на **возвраты товаров или отмененные заказы**. Это подтверждается тем, что все такие записи имеют `TransactionNo`, начинающийся с буквы 'C', что является стандартным обозначением для транзакций отмены или возврата.

**Дальнейшие действия:**

Для целей анализа продаж и выручки, эти записи были **удалены** из основного датафрейма (`df`). Это позволяет сфокусироваться исключительно на данных о состоявшихся продажах, избегая искажения метрик. Если потребуется анализ динамики возвратов, эти данные могут быть выделены и проанализированы отдельно.

---
---


# **Этап 1.5 Решение данных которые вызывают сомнения**

## **Шаг 7 Решение аномалий:**

-----
-----

#**//Общий вывод по первому этапу**

----
----
----

#**Этап 2. Предварительный анализ и создание новых признаков**

- Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
- Создание признаков: месяц, день недели, год из даты.
- Создание признака: общая сумма заказа (`Price` и `Quantity`).
- Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

### **Создание новых признаков из даты**

Для дальнейшего анализа сезонности и временных трендов, из колонки `Date` будут извлечены год, месяц и день недели.

In [33]:
# Извлечение года, месяца и дня недели
df['Year'] = df['Date'].dt.year
df['Month'] = df['Date'].dt.month
df['DayOfWeek'] = df['Date'].dt.dayofweek # Понедельник=0, Воскресенье=6

display(df.head())

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country,Year,Month,DayOfWeek
0,581482,2019-12-09,22485,Set Of 2 Wooden Market Crates,21.47,12,17490,United Kingdom,2019,12,0
1,581475,2019-12-09,22596,Christmas Star Wish List Chalkboard,10.65,36,13069,United Kingdom,2019,12,0
2,581475,2019-12-09,23235,Storage Tin Vintage Leaf,11.53,12,13069,United Kingdom,2019,12,0
3,581475,2019-12-09,23272,Tree T-Light Holder Willie Winkie,10.65,12,13069,United Kingdom,2019,12,0
4,581475,2019-12-09,23239,Set Of 4 Knick Knack Tins Poppies,11.94,6,13069,United Kingdom,2019,12,0


### **Агрегация данных на уровне заказа для анализа корзины**

Для анализа содержимого корзины и среднего чека, агрегируем данные по `TransactionNo`, чтобы получить общую сумму и количество уникальных товаров в каждом заказе.

### **Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен**

Давайте исследуем, какие товары являются наиболее прибыльными и популярными, а также как распределены цены на товары.

### **Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме**

Проанализируем структуру заказов, чтобы понять типичный размер корзины, средний чек и как распределяются заказы по их общей стоимости.

# **Этап 3. Исследовательский анализ данных (EDA)**

*   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
*   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
*   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
*   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
*   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
*   Анализ ценовых сегментов: зависимость количества продаж от цены.

# **Этап 4. Ключевые выводы и рекомендации**
*   Формирование основных инсайтов по каждому блоку анализа.
*   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.

----
----
----

## 📊 SWOT-анализ EDA-проекта

---

### S — Сильные стороны (Strengths)

**Надёжная очистка и подготовка данных**
Данные прошли полную очистку: выполнено преобразование типов (`Date` → `datetime`, `CustomerNo` → `Int64`), обработаны пропуски (55 записей заменены на `-1`), удалены явные дубликаты (5200 записей) и агрегированы неявные (5553 записи). Это обеспечивает высокое качество данных для дальнейшего анализа.

**Эффективная обработка аномалий в количестве товаров**
Отрицательные значения `Quantity` (8548 записей, 1.61% от исходных данных) однозначно идентифицированы как возвраты или отмены и были удалены из основного датафрейма. Это позволяет анализировать только успешные продажи без искажения метрик.

**Богатый и разнообразный датасет**
Очищенный датасет содержит более 522 000 записей, 3753 уникальных товара, 4719 клиентов и транзакции из 38 стран. Такое разнообразие даёт широкие возможности для анализа товаров, клиентов и географии продаж.

**Инсайты по дорогим товарам**
Анализ цен выше 99-го перцентиля выявил товары с аномально высокими ценами (например, «Vintage Red Kitchen Cabinet» за 660.62). Их сохранение позволяет изучать премиальный сегмент и высокомаржинальные позиции.

**Основа для временного анализа**
Извлечение года, месяца и дня недели из даты создаёт базу для анализа трендов, сезонности и покупательского поведения во времени.

---

### W — Слабые стороны (Weaknesses)

**Ограниченные данные о клиентах**
Есть только `CustomerNo`, но нет демографических данных (возраст, пол, доход, история поведения). Это ограничивает возможность глубокой сегментации и персонализированного маркетинга.

**Отсутствие категорий товаров**
Поле `ProductName` содержит свободный текст, нет столбца с категорией. Это затрудняет анализ на уровне групп товаров, выявление слабых категорий и кросс-категорийных покупок.

**Влияние дорогих товаров на метрики**
Наличие товаров с аномально высокой ценой может искажать средний чек и другие показатели. Если такие продажи редки, они создают ложное впечатление о типичном поведении клиентов.

**Неясная природа неявных дубликатов**
Хотя неявные дубликаты были обработаны (суммированием `Quantity`), причина их появления остаётся невыясненной. Это может указывать на проблемы в процессе сбора данных или бизнес-процессах.

**Фокус на выручке, не на прибыли**
Есть цена и количество, что позволяет считать выручку. Но нет данных о себестоимости, поэтому невозможно оценить реальную прибыльность товаров.

**Неясная природа отмен**
Транзакции с префиксом «C» идентифицированы как отмены, но неизвестно, по какой причине они произошли (клиент, склад, оплата). Анализ этих причин мог бы дать инсайты по операционным проблемам.

---

### O — Возможности (Opportunities)

**Глубокая сегментация клиентов (RFM)**
Несмотря на ограниченные данные, можно применить RFM-анализ (давность, частота, сумма) на основе истории транзакций. Это позволит выделить лояльных, «спящих» и новых клиентов для персонализированных кампаний.

**Категоризация товаров**
С помощью текстового анализа или ручной разметки можно добавить категории товаров. Это откроет возможности для управления ассортиментом, кросс-продаж и анализа категорий.

**Монетизация премиальных товаров**
Выявленные дорогие товары указывают на потенциальный нишевый рынок. Можно развивать маркетинг для премиум-сегмента, сотрудничать с дизайнерами или создавать эксклюзивные коллекции.

**Оптимизация цен**
Анализ цены и количества позволяет изучать эластичность спроса и тестировать ценовые стратегии (динамическое ценообразование, пакетные предложения).

**Развитие перспективных рынков**
База клиентов из 38 стран даёт возможность выявить недооценённые регионы и сфокусировать маркетинг, локализовать предложение и улучшить логистику.

**Анализ возвратов**
Хотя возвраты удалены из основного анализа, их можно изучить отдельно. Понимание причин возвратов поможет улучшить качество товаров, описания и снизить возвраты в будущем.

**Оптимизация запасов и поставок**
Анализ популярных и медленно продаваемых товаров позволит лучше управлять запасами, договариваться с поставщиками и запускать акции для неликвидов.

---

### T — Угрозы (Threats)

**Высокая конкуренция в e-commerce**
Рынок товаров для дома и декора очень конкурентен. Без категоризации товаров и глубокого понимания клиентов магазин рискует потерять позиции, особенно на фоне крупных игроков.

**Экономические спады**
Товары для дома — дискреционные покупки. В период нестабильности спрос может резко упасть. Текущие данные не отражают долгосрочные макроэкономические риски.

**Зависимость от одного рынка (например, Великобритания)**
Если основная доля продаж приходится на одну страну, бизнес уязвим к локальным кризисам, изменению законодательства или усилению конкуренции в этом регионе.

**Сбои в цепочке поставок**
Отсутствие данных о поставщиках и категориях делает бизнес уязвимым к перебоям с поставками, дефициту популярных товаров и задержкам.

**Рост стоимости привлечения клиентов**
Без сегментации и программ лояльности затраты на привлечение новых клиентов могут неоправданно расти, особенно в конкурентной среде.

**Изменение трендов**
Тренды в декоре быстро меняются. Без анализа категорий и новых товаров магазин рискует закупить неактуальные позиции или упустить новые тренды.

**Регуляторные риски и доверие клиентов**
Даже при ограниченном объёме данных использование сторонних трекеров и сбор данных требует соблюдения GDPR и других законов. Нарушение может привести к штрафам и репутационным потерям.